include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,377 @@
+                s_apartmentStorage;
+
+            apartment_variable_base()
+            {
+                // TODO: ensure this is a global, investigate cases where address is not reliable due to re-use as stack variables.
+                // printf("%p\n", this);
+            }
+
+            ~apartment_variable_base() = default;
+
+            // non-copyable, non-assignable
+            apartment_variable_base(apartment_variable_base const&) = delete;
+            void operator=(apartment_variable_base const&) = delete;
+
+            // get current value or throw if no value has been set
+            std::any& get_existing()
+            {
+                if (auto any = get_if())
+                {
+                    return *any;
+                }
+                THROW_HR(E_NOT_SET);
+            }
+
+            // get current value or custom-construct one on demand
+            std::any& get_or_create(const std::function<std::any()>& creator)
+            {
+                const auto apartmentId = test_hook::GetApartmentId();
+
+                auto lock = winrt::slim_lock_guard(s_lock);
+
+                auto storage = s_apartmentStorage.find(apartmentId);
+                if (storage != s_apartmentStorage.end())
+                {
+                    auto& variables = storage->second;
+                    auto variable = variables.find(this);
+                    if (variable != variables.end())
+                    {
+                        return variable->second;
+                    }
+                    else
+                    {
+                        variable = variables.insert({ this, creator() }).first;
+                        return variable->second;
+                    }
+                }
+                else
+                {
+                    struct ApartmentObserver : public winrt::implements<ApartmentObserver, IApartmentShutdown>
+                    {
+                        void STDMETHODCALLTYPE OnUninitialize(unsigned long long apartmentId) noexcept override
+                        {
+                            // This code runs at apartment rundown so be careful to avoid deadlocks by
+                            // extracting the variables under the lock then release them outside.
+                            auto variables = [apartmentId]()
+                            {
+                                auto lock = winrt::slim_lock_guard(s_lock);
+                                return s_apartmentStorage.extract(apartmentId);
+                            }();
+                        }
+                    };
+
+                    test_hook::RegisterForApartmentShutdown(winrt::make<ApartmentObserver>().get());
+
+                    auto variables = s_apartmentStorage.insert({ apartmentId,
+                        { { this, creator() } } }).first;
+                    return variables->second.begin()->second; // new map entry is first
+                }
+            }

```

Comment:
Need to call the `creator()` outside the lock to avoid reentrancy deadlock. Can also factor out the "create outside the lock" code as well as the "find current apartment" code.

```cpp
            using variables_map = std::unordered_map<wil::details::apartment_variable_base<test_hook>*, std::any>>;
            variables_map* get_current_apartment_variables()
            {
                const auto apartmentId = test_hook::GetApartmentId();

                auto storage = s_apartmentStorage.find(apartmentId);

                if (storage != s_apartmentStorage.end())
                {
                    return &storage->second;
                }
                return nullptr;
            }

            variables_map* ensure_current_apartment_variables()
            {
                auto variables = get_current_apartment_variables();
                if (variables)
                {
                    return variables;
                }

                struct ApartmentObserver : public winrt::implements<ApartmentObserver, IApartmentShutdown>
                {
                    void STDMETHODCALLTYPE OnUninitialize(unsigned long long apartmentId) noexcept override
                    {
                        // This code runs at apartment rundown so be careful to avoid deadlocks by
                        // extracting the variables under the lock then release them outside.
                        auto variables = [apartmentId]()
                        {
                            auto lock = winrt::slim_lock_guard(s_lock);
                            return s_apartmentStorage.extract(apartmentId);
                        }();
                    }
                };
                test_hook::RegisterForApartmentShutdown(winrt::make<ApartmentObserver>().get());

                return &s_apartmentStorage.insert({ apartmentId, {} }).first->second;
            }

            std::any& get_or_create(const std::function<std::any()>& creator)
            {
                variables_map* variables;

                { // scope for lock
                    auto lock = winrt::slim_lock_guard(s_lock);
                    variables = ensure_current_apartment_variables();

                    auto variable = variables->find(this);
                    if (variable != variables->end())
                    {
                        return variable->second;
                    }
                } // drop the lock

                // create the object outside the lock to avoid reentrancy deadlock
                auto value = creator();

                auto insert_lock = winrt::slim_lock_guard(s_lock);
                // The insertion may fail if creator() recursively caused itself to be created,
                // in which case we return the existing object and the falsely-created one is discarded.
                return variables->insert({ this, std::move(value) }).first->first;
            }
```

Reply:
thanks for the impl. updated